src/librustc_typeck/check/expr.rs

```diff
@@ -1509,13 +1509,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         self.tcx().ty_error()
     }
 
+    fn suggest_await_on_field_access(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        field_ident: Ident,
+        base: &'tcx hir::Expr<'tcx>,
+        expr: &'tcx hir::Expr<'tcx>,
+        def_id: DefId,
+    ) {
+        let param_env = self.tcx().param_env(def_id);
+        let future_trait = self.tcx.require_lang_item(LangItem::Future, None);
+        // Future::Output
+        let item_def_id =
+            self.tcx.associated_items(future_trait).in_definition_order().next().unwrap().def_id;
+
+        let mut projection_ty = None;
+        for (predicate, _) in self.tcx.predicates_of(def_id).predicates {
+            if let ty::PredicateAtom::Projection(projection_predicate) = predicate.skip_binders() {
+                if item_def_id == projection_predicate.projection_ty.item_def_id {
+                    projection_ty = Some(projection_predicate.projection_ty);
+                    break;
+                }
+            }
+        }
+        debug!("suggest_await_on_field_access: projection_ty={:?}", projection_ty);
+
+        let cause = self.misc(expr.span);
+        let mut selcx = SelectionContext::new(&self.infcx);
+
+        let mut obligations = vec![];
+        if let Some(projection_ty) = projection_ty {
+            let normalized_ty = rustc_trait_selection::traits::normalize_projection_type(
+                &mut selcx,
+                param_env,
+                projection_ty,
+                cause,
+                0,
+                &mut obligations,
+            );
```

Comment:
Is there a central place somewhere that we could define a helper to do this? I'd rather not duplicate it in all these places.

(If we can't extract the normalizing code that's fine, I'd prefer to, but we could just extract the parts before it.)

Reply:
Having a `src/librustc_typeck/check/diagnostics.rs` would make sense, and would follow the pattern in other parts of the compiler.

I made a quick search for `SelectionContext::new(` as I would imagine that any code doing this same thing would be using it, but I couldn't find any _diagnostics_ use in `typeck`.